### PR TITLE
Handling out of memory errors gracefully 

### PR DESCRIPTION
**Pull request by SherineAwad**
Working on issue #510

### CODE REVIEW

**Comment by ctb** [comment]
@SherineAwad, what is this PR about? Could you edit the title and description, please?

**Comment by SherineAwad** [comment]
@ctb I am working on this one https://github.com/ged-lab/khmer/issues/510

**Comment by ctb** [comment]
OK - a few requests --

please edit the title (currently "testing") to say "Handling out of memory errors gracefully."

please edit the description (which currently says "work in progress") to say "Working on issue #510"

thanks!

**Comment by SherineAwad** [comment]
Done !

*Sherine Awad *
Postdoctoral Research Fellow
Lab of Genomics, Evolution and Development
Computer Science Dept. - Michigan State University

**Comment by SherineAwad** [comment]
@mr-c  What I have done so far is catching bad allocs.

File: khmer/_khmermodule.cc [hunk 1]
```diff
@@ -1628,8 +1629,10 @@ static PyObject* new_hashtable(PyObject * self, PyObject * args)
     if (kcounting_obj == NULL) {
         return NULL;
     }
-
-    kcounting_obj->counting = new CountingHash(k, size);
+    try {
+	kcounting_obj->counting = new CountingHash(k, size);}
+    catch (const std::bad_alloc&) {
+    	exit(EXIT_FAILURE);}
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: at the minimum, use `PyErr_SetString(PyExc_MemoryError)` instead of exit.

**Comment by SherineAwad** [comment]
@mr-c  Ready for review

- [x] Is it mergable?
- [x] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage.
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

File: ChangeLog [hunk 2]
```diff
@@ -1,3 +1,7 @@
+2015-02-19 Sherine Awad <sherine.awad@gmail.com>
+
+  * khmer/_khmermodule.cc: Catch out of memory errors gracefully.
```
Comment (mr-c) [inline ChangeLog]: The bullet point should line up with the others; one more space please

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,7 @@
+2015-02-19 Sherine Awad <sherine.awad@gmail.com>
```
Comment (mr-c) [inline ChangeLog]: Two spaces between your name and both the date and your email address.

File: khmer/_khmermodule.cc [hunk 4]
```diff
@@ -1,5 +1,5 @@
 //
-// This file is part of khmer, http://github.com/ged-lab/khmer/, and is
+// Thit file is part of khmer, http://github.com/ged-lab/khmer/, and is
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: ?

File: khmer/_khmermodule.cc [hunk 5]
```diff
@@ -23,7 +23,7 @@
 #include "labelhash.hh"
 #include "khmer_exception.hh"
 #include "hllcounter.hh"
-
+//#include <stdlib.h>
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: Please undo this change. Try to check your pull requests before requesting review either via the GitHub interface ( https://github.com/ged-lab/khmer/pull/771/files ) or via `git fetch; git diff origin/master`

File: khmer/_khmermodule.cc [hunk 6]
```diff
@@ -1348,8 +1348,8 @@ static PyObject * hash_get_tags_and_positions(PyObject * self, PyObject * args)
     while (!kmers.done()) {
         HashIntoType kmer = kmers.next();
         if (set_contains(counting->all_tags, kmer)) {
-             posns.push_back(pos);
-             tags.push_back(kmer);
+            posns.push_back(pos);
+            tags.push_back(kmer);
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: Is this from running `make format`?

File: khmer/_khmermodule.cc [hunk 7]
```diff
@@ -1628,8 +1628,13 @@ static PyObject* new_hashtable(PyObject * self, PyObject * args)
     if (kcounting_obj == NULL) {
         return NULL;
     }
+    try {
+        kcounting_obj->counting = new CountingHash(k, size);
+    } catch (const std::bad_alloc&) {
+        PyErr_SetString(PyExc_MemoryError,"Out of Memory");
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: Is the string necessary? I thought `PyExc_MemoryError` had a default message.

File: khmer/_khmermodule.cc (hunk 6 above)
Comment (SherineAwad) [inline khmer/_khmermodule.cc]: yes

File: khmer/_khmermodule.cc [hunk 8]
```diff
@@ -3422,7 +3432,15 @@ static PyObject* khmer_hashbits_new(PyTypeObject * type, PyObject * args,
...
 
-        self->hashbits = new Hashbits(k, sizes);
+        try {
+            self->hashbits = new Hashbits(k, sizes);
+        } catch (const std::bad_alloc&) {
+            PyErr_SetString(PyExc_MemoryError,"Out of Memory");
+            return NULL;
+        }
+
+
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: Too many empty lines here

**Comment by mr-c** [comment]
Good start, thanks.

File: khmer/_khmermodule.cc (hunk 7 above)
Comment (SherineAwad) [inline khmer/_khmermodule.cc]: Yes it is. I t requires two arguments.

Comment (mr-c) [inline khmer/_khmermodule.cc]: Ah, I was thinking of `PyErr_SetNone(PyExc_MemoryError);` Turns out this was because I had already done much of this already; see: https://github.com/ged-lab/khmer/issues/510#issuecomment-75154752

**Comment by SherineAwad** [comment]
@mr-c  So should I close this PR if it is handled in #704 ?

**Comment by mr-c** [comment]
Yes, please
